t/4: Added tests for "utils"

fredck opened the pull request:
Fixes #4.

This PR follows depends on https://github.com/ckeditor/ckeditor5/issues/1.

gregpabian commented on tests/utils/utils.js:
```diff
@@ -0,0 +1,153 @@
+
+'use strict';
+
+describe( 'extend()', function() {
```
I'd wrap all of these describes with one more:
```
describe('Utils', ...
```

Then it would produce clearer results, like:
"Utils extend() should extend and override"

fredck replied:
Do we really need to make the file more bloated, considering that we're already inside the "utils" test page?

Then, the correct name is "utils" (lowercased) and this would make things ackward because there is no dot between "utils" and "extend()": 
"utils extend() should extend and override"

So I think that we should either change the current describe to "utils.extend()" or leave it as it is  now (my option).

gregpabian commented on tests/utils/utils.js:
```diff
@@ -0,0 +1,153 @@
+			expect( utils.isObject( Date.now() ) ).to.be.false();
+			expect( utils.isObject( f1 ) ).to.be.false();
+			expect( utils.isObject( f2 ) ).to.be.false();
+
```
I think you should add a test for `null` :trollface:

fredck replied:
True :) just pushed it.

gregpabian commented:
I've been thinking about all those `CKEDITOR.require` calls and in this case we can use mocha's hooks to simplify the code:
```js
describe( 'extend()', function() {
	var utils;

	// hooks can be asynchronous so we can do the following
	beforeEach( function( done ) {
		CKEDITOR.require( [ 'utils' ], function( u ) {
			utils = u;
			done();
		} );
	} );

	// not async so no need for done callback, yay!
	it( 'should extend and override', function() {
		var target = {
			a: 1,
			b: 2
		};

		var extensions = {
			a: 'A',
			c: 3
		};

		var ret = utils.extend( target, extensions );

		expect( target.a ).to.equal( 'A' );
		...
	} );
} );
```

Besides, I don't think we have to require the library for every test, so we can replace `beforeEach()` with `before()`.

What do you think about this?

Reinmar commented:
> What do you think about this?

Looks good, but I wonder if we can simplify this even further by creating some helper. It could work this way:

```
var modules;

before( function( done ) {
    doSomeStuffThatYouNeed();
    bender.amd.require( done, modules, [ 'Promise', 'utils' ] );
} );

// modules.Promise;
// modules.utils;
```

And even simpler, if one does not need the `before()` callback:

```
var modules = bender.amd.require( [ 'Promise', 'utils' ] );
```

And the `require()` function, if only one argument was passed, would use `before()` internally.

Reinmar commented:
One more thing that we'll need to understand - how can we inject some dependencies by overriding what `CKEDITOR.require` returns when loading a module. Most likely it could look like this:

```
it( 'test Foo with mocked Bar', function() {
    bender.amd.inject( 'Bar', ourMockModule );

   CKEDITOR.require( 'Foo', function( Foo ) {
      // Foo internally requires Bar and since we injected ourMockModule
      // it should now use our mock instead of the real Bar.
      done(); 
   } );
} );
```

And the injection should be reverted in `afterEach/beforeEach`, so this TC does not affect the next one.

Does it make sense?

PS. I guess there must be some library that has these features :D.

gregpabian commented:
> bender.amd.require( done, modules, [ 'Promise', 'utils' ] );

You couldn't think of a worse argument order, could you? :trollface:

Reinmar commented:
> You couldn't think of a worse argument order, could you?

My apologies :D. I was changing it too many times. Actually, to unify with the 2nd version I would do:

```
modules = bender.amd.require( [ 'Promise', 'utils' ], done );
```

gregpabian commented:
> how can we inject some dependencies

I think the most common way is to use [Require's mappings](http://requirejs.org/docs/api.html#config-map), but if we decide to come out with our own helper for this purpose, then I'm OK with that.

fredck commented:
What do you guys think about this?

```
bender.injectModules( 'ckeditor', 'utils' );

describe( 'test', function( ckeditor, utils ) {
    ...
}
```

Basically, the ckeditor5 plugin for bender will add `injectModules` and override `describe` to have them as parameters.

Reinmar commented:
> I think the most common way is to use Require's mappings, but if we decide to come out with our own helper for this purpose, then I'm OK with that.

I only briefly scanned the mappings doc, but it does not look like a big help, because it'd need to be per file. I would love to have something:

1. Automated.
2. Local.
3. Temporary.

Reinmar commented:
> Basically, the ckeditor5 plugin for bender will add injectModules and override describe to have them as parameters.

The API looks cool. If it's possible to code it this way (`injectModules` is outside of `describe`), then it'd be a way to go. The other thing is the method name - injecting relates to dependency injection and that's the second feature we need, so I would simply call it `bender.amd.require()`.

fredck commented:
@Reinmar's [comment](https://github.com/ckeditor/ckeditor5-core/pull/5#issuecomment-70990318):

> Temporary.

I think it is not possible.

Once RequireJS returns `Foo`, from your example, it'll always return that same object. It'll not "recreate" it every time require() is called, so you cannot replace `Bar` inside it so dynamically.

We may figure out a way to make this module override happen once for the whole test page... or even check if there is not really a way to force RequireJS to "forget" the previous module definition and recreate it.

fredck commented:
> so I would simply call it bender.amd.require().

Cool, I'll give it a try and possibly update the PRs.

fredck commented:
Btw guys, we're dirtying up this PR too much with side talks. The discussion about module overriding should go somewhere else in an issue.

fredck commented:
Just pushed `bender.amd.require()` in [the relative ckeditor5 PR](https://github.com/ckeditor/ckeditor5/pull/11), updating this one accordingly.

You gonna love this :) (I hope!)

fredck commented:
Ok, updated ckeditor5 with the new strategy for `bender.amd.require()` and updated the tests accordingly.

Now the remaining topic... as you can see, every `describe()` in the same test page has its own call to `bender.amd.require()`, repetitively. This is done to not pollute the global namespace and to avoid making things uglier with a big closure.

Options:

 1. Leave it this way.
 2. Pollute the global namespace with this one and only variable named `modules`.
 3. Use a big and ugly closure only when we have several describe()s that share the same modules.
 4. Use a big and ugly closure always.

I would be definitely ok with "2" as it is simple and it should not be a big deal to have the global a bit dirty (but under control) in our tests.

Reinmar commented:
AFAIR Bender wraps a test file with a closure so the global namespace should not be polluted. But even if not, I'm still ok with 2.

gregpabian commented:
> Bender wraps a test file with a closure so the global namespace should not be polluted

That's true - the test script lands in a closure so it won't be a problem.

fredck commented:
> That's true - the test script lands in a closure so it won't be a problem.

Wonderful! Just changed this. We should have the final version ready for review now.

gregpabian commented on tests/utils/utils.js:
```diff
@@ -0,0 +1,146 @@
+		expect( utils.extend( target, true ) ).to.equal( target );
+		expect( utils.extend( target, undefined ) ).to.equal( target );
+		expect( utils.extend( target, [] ) ).to.equal( target );
+		expect( utils.extend( target, Date.now() ) ).to.equal( target );
```
`Date.now()` returns a number, so what's the point of this case since the number is being tested few lines earlier?

gregpabian commented on tests/utils/utils.js:
```diff
@@ -0,0 +1,146 @@
+		expect( utils.isFunction( undefined ) ).to.be.false();
+		expect( utils.isFunction( [] ) ).to.be.false();
+		expect( utils.isFunction( {} ) ).to.be.false();
+		expect( utils.isFunction( Date.now() ) ).to.be.false();
```
Again, you're testing a number which has already been tested few lines earlier.

fredck commented:
@gregpabian, I've fixed your last comments. My mistake.

Reinmar commented:
> Once RequireJS returns Foo, from your example, it'll always return that same object. It'll not "recreate" it every time require() is called, so you cannot replace Bar inside it so dynamically.

Just found this:

> There is a global function, requirejs.undef(), that allows undefining a module. It will reset the loader's internal state to forget about the previous definition of the module.

http://requirejs.org/docs/api.html#undef
